### PR TITLE
 Modify population dislocation to have choice dislocation and unsafe occupancy 

### PR DESCRIPTION
**Pull request by Vismayak**
Adding the requested changes from Nathanael to have flags that calculate choice_dislocation and unsafe occupancy

### CODE REVIEW

**Comment by Vismayak** [comment]
Contacted Nathanael and I believe the last step was updating dislocation depending on hoice dislocation and unsafe occupancy. I think last step would be figuring out whether to modify semantics for current popdislocation dataframe or create a different version

File: pyincore/analyses/populationdislocation/populationdislocation.py [hunk 1]
```diff
@@ -116,12 +152,28 @@ def run(self):
...
         merged_block_inv = PopulationDislocationUtil.merge_damage_housing_block(
             building_dmg, housing_unit_alloc, bg_data
         )
 
         # Returns dataframe
         merged_final_inv = self.get_dislocation(seed_i, merged_block_inv, value_loss)
+
+        # Choice dislocation and unsafe occupancy calculations
+        if choice_dislocation:
+            choice_dislocation_cutoff = self.get_parameter("choice_dislocation_cutoff") or 0.5
```
Comment (longshuicy) [inline pyincore/analyses/populationdislocation/populationdislocation.py:168]: Neat. I like this one liner way of fallback to default.

**Review by longshuicy — CHANGES_REQUESTED** [review]

File: pyincore/analyses/populationdislocation/populationdislocationutil.py [hunk 2]
```diff
@@ -179,3 +179,47 @@ def compare_columns(table, col1, col2, drop):
...
+            null: None.
+
+        """
+        condition1 = pop_dislocation['dislocated'] == 1
+        condition2 = pop_dislocation[choice_dislocation_ds] > choice_dislocation_cutoff
+        pop_dislocation['choice_dislocation'] = np.where(condition1 & condition2, 1, 0)
+        # Change dislocated to 0 if choice dislocation is 1
+        pop_dislocation['dislocated'] = np.where(condition1 & condition2, 0, 1)
+
+        return None
```
[the remark this answers is not in the document]
Comment (Vismayak) [reply pyincore/analyses/populationdislocation/populationdislocationutil.py:203]: I think the data frame is passed on by reference so changes will show but I will double check

[the remark this answers is not in the document]
Comment (Vismayak) [reply pyincore/analyses/populationdislocation/populationdislocationutil.py:203]: image
In the test choice_dislocation is coming in the final result so I think it works

File: pyincore/analyses/populationdislocation/populationdislocationutil.py [hunk 3]
```diff
@@ -179,3 +179,47 @@ def compare_columns(table, col1, col2, drop):
                 table = table.drop(columns=[col1, col2, col1 + "-" + col2])
 
         return table
+
+
+    @staticmethod
+    def get_choice_dislocation(pop_dislocation, choice_dislocation_cutoff, choice_dislocation_ds):
```
Comment (longshuicy) [inline pyincore/analyses/populationdislocation/populationdislocationutil.py:185]: I think pep8 is complaining too many white lines

**Comment by navarroc** [comment]
Should we always add all the new dislocation columns even if the new dislocation isn't selected (but empty)? @longshuicy @Vismayak ?

**Comment by Vismayak** [comment]
@navarroc could you please check the PR again, I have added the columns to be permanent even if flags are not called and to have boolean values.

**Comment by navarroc** [comment]
Should unsafe_occupancy be an empty column instead of initialized to False if we are not setting choice_dislocation? 

Also, we may want to trim the column name to unsafe_occ so it fits into 10 character limit if joining with a shapefile.

**Comment by navarroc** [comment]
Sorry, my bad, the test is setting unsafe_occupancy value, ignore my first statement, but the length of the column is still a question

**Comment by navarroc** [comment]
We may also want to shorten choice_dislocation to choice_dis so it meets the 10 character limit for shapefiles.

File: pyincore/analyses/populationdislocation/populationdislocationutil.py [hunk 4]
```diff
@@ -207,7 +207,7 @@ def get_unsafe_occupancy(pop_dislocation, unsafe_occupancy_cutoff, unsafe_occupa
         """Get unsafe occupancy based on the dislocation state and the unsafe occupancy cutoff.
 
         Args:
-            pop_dislocation (np.array): Population dislocation Result.
+            pop_disloncation (np.array): Population dislocation Result.
```
Comment (navarroc) [inline pyincore/analyses/populationdislocation/populationdislocationutil.py:210]: can you fix the typo is dislocation?

**Review by longshuicy — APPROVED** [review]
Looks good.

File: pyincore/analyses/populationdislocation/populationdislocation.py [hunk 5]
```diff
@@ -116,12 +152,30 @@ def run(self):
...
+            choice_dislocation_ds = self.get_parameter("choice_dislocation_ds") or "DS_0"
+            PopulationDislocationUtil.get_choice_dislocation(merged_final_inv, choice_dislocation_cutoff, choice_dislocation_ds)
+
+        if unsafe_occupancy:
+            unsafe_occupancy_cutoff = self.get_parameter("unsafe_occupancy_cutoff") or 0.5
+            unsafe_occupancy_ds = self.get_parameter("unsafe_occupancy_ds") or "DS_3"
+            PopulationDislocationUtil.get_unsafe_occupancy(merged_final_inv, unsafe_occupancy_cutoff, unsafe_occupancy_ds)
+
         csv_source = "dataframe"
```
Comment (navarroc) [inline pyincore/analyses/populationdislocation/populationdislocation.py:179]: unused variable

File: pyincore/analyses/populationdislocation/populationdislocationutil.py [hunk 6]
```diff
@@ -179,3 +179,46 @@ def compare_columns(table, col1, col2, drop):
...
+        Args:
+            pop_dislocation (np.array): Population dislocation Result.
+            unsafe_occupancy_cutoff (float): Unsafe occupancy cutoff.
+            unsafe_occupancy_ds (int): Unsafe occupancy damage state.
+
+        Returns:
+            null: None.
+
+        """
+        condition1 = pop_dislocation['dislocated'] == False
```
Comment (navarroc) [inline pyincore/analyses/populationdislocation/populationdislocationutil.py:217]: pep8 doesn't seem to like this. Is this setting the value to False if pop_dislocation['dislocated'] is empty?

Comment (Vismayak) [reply pyincore/analyses/populationdislocation/populationdislocationutil.py:217]: It did not give the error when I id `== 0`, I can revert back to that. Nothing else comes to mind

File: CHANGELOG.md [hunk 7]
```diff
@@ -6,6 +6,11 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/)
 and this project adheres to [Semantic Versioning](http://semver.org/).
 
 
+## [Unreleased]
+
+### Added
+- Add support for choice disloaction and unsafe occupancy in the population dislocation analysis [#525](https://github.com/IN-CORE/pyincore/issues/525)
```
Comment (navarroc) [inline CHANGELOG.md:12]: last change, can you fix the typo here :)

**Review by navarroc — APPROVED** [review]

**Review by longshuicy — APPROVED** [review]

**Pull request merged**